File: android/besitos-android-sdk/src/main/kotlin/io/github/kashkick/besitos/android/sdk/offerwall/validation/ValidationUtil.kt
package io.github.kashkick.besitos.android.sdk.offerwall.validation

import io.github.kashkick.besitos.android.sdk.offerwall.config.OfferwallConfig

internal object ValidationUtil {

    private val SAFE_PATTERN = Regex("^[A-Za-z0-9\\-_]*$")

    fun validate(config: OfferwallConfig): ValidationResult {
        if (config.partnerId.isBlank()) {
            return ValidationResult.Invalid("partnerId is required and cannot be blank")
        }
        if (!SAFE_PATTERN.matches(config.partnerId)) {
            return ValidationResult.Invalid("partnerId contains invalid characters")
        }
        if (config.userId.isBlank()) {
            return ValidationResult.Invalid("userId is required and cannot be blank")
        }
        if (!SAFE_PATTERN.matches(config.userId)) {
            return ValidationResult.Invalid("userId contains invalid characters")
        }
        config.subId?.takeIf { it.isNotBlank() }?.let {
            if (!SAFE_PATTERN.matches(it)) return ValidationResult.Invalid("subId contains invalid characters")
        }
        config.deviceId?.takeIf { it.isNotBlank() }?.let {
            if (!SAFE_PATTERN.matches(it)) return ValidationResult.Invalid("deviceId contains invalid characters")
        }
        config.idfa?.takeIf { it.isNotBlank() }?.let {
            if (!SAFE_PATTERN.matches(it)) return ValidationResult.Invalid("idfa contains invalid characters")
        }
        config.gdfa?.takeIf { it.isNotBlank() }?.let {
            if (!SAFE_PATTERN.matches(it)) return ValidationResult.Invalid("gdfa contains invalid characters")
        }
        return ValidationResult.Valid
    }
}

sealed class ValidationResult {
    data object Valid : ValidationResult()
    data class Invalid(val reason: String) : ValidationResult()
}

File: ios/Sources/Offerwall/ValidationUtil.swift
import Foundation

enum ValidationResult {
    case valid
    case invalid(String)
}

struct ValidationUtil {
    private static let safePattern = "^[A-Za-z0-9\\-_]*$"

    static func validate(_ config: OfferwallConfig) -> ValidationResult {
        guard !config.partnerId.isEmpty else {
            return .invalid("partnerId is required and cannot be empty")
        }
        guard isSafe(config.partnerId) else {
            return .invalid("partnerId contains invalid characters")
        }
        guard !config.userId.isEmpty else {
            return .invalid("userId is required and cannot be empty")
        }
        guard isSafe(config.userId) else {
            return .invalid("userId contains invalid characters")
        }
        if let subId = config.subId, !subId.isEmpty, !isSafe(subId) {
            return .invalid("subId contains invalid characters")
        }
        if let deviceId = config.deviceId, !deviceId.isEmpty, !isSafe(deviceId) {
            return .invalid("deviceId contains invalid characters")
        }
        if let idfa = config.idfa, !idfa.isEmpty, !isSafe(idfa) {
            return .invalid("idfa contains invalid characters")
        }
        if let gdfa = config.gdfa, !gdfa.isEmpty, !isSafe(gdfa) {
            return .invalid("gdfa contains invalid characters")
        }
        return .valid
    }

    private static func isSafe(_ value: String) -> Bool {
        let regex = try? NSRegularExpression(pattern: safePattern)
        let range = NSRange(value.startIndex..., in: value)
        return regex?.firstMatch(in: value, range: range) != nil
    }
}
